src/schema/search.ts

```diff
@@ -0,0 +1,155 @@
+      value: "Artwork",
+    },
+    ARTIST: {
+      value: "Artist",
+    },
+  },
+})
+
+export const SearchMode = new GraphQLEnumType({
+  name: "SearchMode",
+  values: {
+    AUTOSUGGEST: {
+      value: "AUTOSUGGEST",
+    },
+    SITE: {
+      value: "SITE",
+    },
+  },
+})
+
+export const searchArgs = pageable({
+  query: {
+    type: new GraphQLNonNull(GraphQLString),
+    description: "Search query to perform. Required.",
+  },
+  indexes: {
+    type: new GraphQLList(SearchIndex),
+    description: "Indexes to include in search. Default: [ARTIST, ARTWORK].",
+  },
+  mode: {
+    type: SearchMode,
+    description: "Mode of search to exceute. Default: SITE.",
+  },
+})
+
+const hydrateSearchResult = (result, root) => {
+  const loaderMapping = {
+    Artist: root.artistLoader,
+    Artwork: root.artworkLoader,
+  }
+
+  const loader = loaderMapping[result.label]
+
+  return loader(result.id).then(object => {
+    return {
+      ...object,
+      __typename: result.label,
+    }
+  })
+}
+
+const shouldHydrate = info => {
+  let shouldHydrate = false
+
+  visit(info.fieldNodes[0], {
+    Field(node) {
+      if (node.name.value === "node") {
+        visit(node, {
+          InlineFragment(node) {
+            if (
+              node.typeCondition &&
+              node.typeCondition.name.value !== SearchableItem.name
+            ) {
+              shouldHydrate = true
+              return BREAK
+            }
+          },
+          FragmentSpread(_node) {
+            throw new Error(
+              "Named fragment spreads are currently unsupported for search."
+            )
+          },
+        })
+      }
+    },
+  })
+
+  return shouldHydrate
+}
+
+const SearchConnection = connectionWithCursorInfo(Searchable)
+
+export const Search: GraphQLFieldConfig<any, any, any> = {
+  type: SearchConnection,
+  description: "Global search",
+  args: searchArgs,
+  resolve: (source, args, _request, info) => {
+    const pageOptions = convertConnectionArgsToGravityArgs(args)
+
+    const gravityArgs = {
+      ...pageOptions,
+      indexes: args.indexes,
+      total_count: true,
+    }
+
+    const hydrate = shouldHydrate(info)
+
+    return source.searchLoader(gravityArgs).then(({ body, headers }) => {
+      const totalCount = parseInt(headers["x-total-count"])
+      const pageCursors = createPageCursors(pageOptions, totalCount)
+      // args.last = pageCursors.last.page
+
+      return Promise.all(
+        body.map(result => {
+          if (hydrate) {
+            return hydrateSearchResult(result, source)
+          } else {
+            return Promise.resolve({
+              ...result,
+              __typename: "SearchableItem",
+            })
+          }
+        })
+      ).then(results => {
+        const connection = connectionFromArraySlice(results, args, {
+          arrayLength: totalCount,
+          sliceStart: pageOptions.offset,
+        })
+        // TODO: We shouldn't have to override the pageInfo here
```

Comment:
I don't fully follow the comment, but I think you do need to override/merge the modified implementations for `hasPreviousPage` and `hasNextPage`, so this looks good to me.

The way those fields are calculated by the lib we're using doesn't fully work. Here's that implementation: https://github.com/graphql/graphql-relay-js/blob/582b0162709bd2db45cbd3925e2e69d9d269ec7e/src/connection/arrayconnection.js#L106-L108 

As you're scrolling forward ', by using `(first: ..., after: ...)` as your connection paging arguments, it can't determine if there are previous pages (and vice versa), so the lib returns `false` (kind of a weird return). So this slightly laborious construction of the connection return type (cursors, count, connection, page info), seems 👍 . We do this in several places so it's worth refactoring or extracting that into a wrapper of some kind.

Reply:
When we stepped into the lib code it seemed mostly due to a missing `last` (page number) parameter, but for some reason the code that determined `hasPreviousPage` would still result in `false` for reasons that weren't immediately clear. In short, this is a comment to look into that again at some point and clean all these cases up, as it should be able to work if we pass in the right parameters.

Reply:
Well, those are missing b/c that's actually passed in by the caller. You have two options for scrolling: 'forward' which means providing `(first: ..., after: ....`), or 'backward' which means providing `(last: ...., before: ...)`. Theoretically those are mutually exclusive (but we don't do that validation).

Since we just do regular `page/size` and windowed pagination, we _always_ paginate those via `(first: ..., after: ...)`, and so that's what requires the manual `hasXPage` calculation.